Sources/PrivacyStats/internal/PrivacyStatsUtils.swift

```diff
@@ -0,0 +1,118 @@
+        }
+        return statsObjects
+    }
+
+    /**
+     * Returns a dictionary representation of blocked trackers counts grouped by company name for the current day.
+     */
+    static func loadCurrentDayStats(in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let startDate = Date.currentPrivacyStatsPackTimestamp
+        return try loadBlockedTrackersStats(since: startDate, in: context)
+    }
+
+    /**
+     * Returns a dictionary representation of blocked trackers counts grouped by company name for past 7 days.
+     */
+    static func load7DayStats(in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let startDate = Date().privacyStatsOldestPackTimestamp
+        return try loadBlockedTrackersStats(since: startDate, in: context)
+    }
+
+    private static func loadBlockedTrackersStats(since startDate: Date, in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let request = NSFetchRequest<NSFetchRequestResult>(entityName: "DailyBlockedTrackersEntity")
+        request.predicate = NSPredicate(format: "%K >= %@", #keyPath(DailyBlockedTrackersEntity.timestamp), startDate as NSDate)
+
+        let companyNameKey = #keyPath(DailyBlockedTrackersEntity.companyName)
+
+        // Expression description for the sum of count
+        let countExpression = NSExpression(forKeyPath: #keyPath(DailyBlockedTrackersEntity.count))
+        let sumExpression = NSExpression(forFunction: "sum:", arguments: [countExpression])
+
+        let sumExpressionDescription = NSExpressionDescription()
+        sumExpressionDescription.name = "totalCount"
+        sumExpressionDescription.expression = sumExpression
+        sumExpressionDescription.expressionResultType = .integer64AttributeType
+
+        request.propertiesToGroupBy = [companyNameKey]
+        request.propertiesToFetch = [companyNameKey, sumExpressionDescription]
+        request.resultType = .dictionaryResultType
+
+        let results = (try context.fetch(request) as? [[String: Any]]) ?? []
+
+        let groupedResults = results.reduce(into: [String: Int64]()) { partialResult, result in
+            if let companyName = result[companyNameKey] as? String, let totalCount = result["totalCount"] as? Int64, totalCount > 0 {
+                partialResult[companyName] = totalCount
+            }
+        }
+
+        return groupedResults
+    }
+
+    /**
+     * Deletes stats older than 7 days for all companies.
+     */
+    static func deleteOutdatedPacks(in context: NSManagedObjectContext) {
+        let oldestValidTimestamp = Date().privacyStatsOldestPackTimestamp
+
+        let request = DailyBlockedTrackersEntity.fetchRequest()
+        request.predicate = NSPredicate(format: "%K < %@", #keyPath(DailyBlockedTrackersEntity.timestamp), oldestValidTimestamp as NSDate)
+        context.deleteAll(matching: request)
```

Comment:
Eventually this calls a `NSManagedObjectContext` extension which loops over entities and deletes them all. Would it be more efficient to use a `NSBatchDeleteRequest` here? e.g:

```
let request = NSBatchDeleteRequest(fetchRequest: DailyBlockedTrackersEntity.fetchRequest())
request.predicate = NSPredicate(format: "%K < %@", Keys.timestamp, oldestValidTimestamp as NSDate)
try context.execute(request)
```

Reply:
Great idea, tbf I forgot NSBatchDeleteRequest existed :) updated it now and verified with tests that it works well!